Add tokenStart and tokenEnd to query

North101 opened the pull request:
I wanted to be able to offer auto complete suggestions for my search query and for that I needed to know the location of the `Query` to figure out if it was at my cursor position.

This PR adds the start / end position of the `Query` relative to `input`.

North101 commented:
I also would like to make the `<exp>` part of `field:<exp>` and `field=<exp>` optional. Though this is would be a breaking change. If you don't want that I'm happy to just keep it in my fork

isoos commented on lib/src/ast.dart:
```diff
@@ -1,5 +1,13 @@
+  const Query(this.tokenStart, this.tokenEnd);
+
+  /// The start position of the token from the input.
+  final int tokenStart;
```
This name feels a bit strange, how about `startIndex` instead? also `endIndex`?
Or event better, following the Dart SDK's `Match` class using `start` and `end`, with similar semantics:
https://api.dart.dev/stable/2.17.5/dart-core/Match/end.html

North101 replied:
I originally had it as `start` and `end` but that clashed with `RangeQuery`. I think `startIndex` and `endIndex` are better names

isoos reviewed (changes requested):
I think this is a good feature to add, but please:
- add tests, the more the better to see (a) what is now available and (b) how to use/access it (it can be a new file)
- explain a bit how the optional `exp` is breaking, and maybe also add tests there too with comments
- also update the changelog and bump the version

North101 commented:
> * add tests, the more the better to see (a) what is now available and (b) how to use/access it (it can be a new file)

Sure. I didn't want to write them until you gave it an ok :P

> * explain a bit how the optional `exp` is breaking, and maybe also add tests there too with comments

It was failing one test but I was able to make a minor change `or` to get it to pass. Another reason it might be breaking is because the following used to error but now parses because it used to expect something after the `:`:
```
field: AND somethingelse
```

> * also update the changelog and bump the version

Cool. Will do after the tests

North101 commented:
I've changed `FieldScopeQuery` and `FieldCompareQuery` to take `TextQuery` instead of `String` as I found I needed to know the start/end index of the `field`. I added a getter to keep them backwards compatible.

isoos commented:
> I've changed FieldScopeQuery and FieldCompareQuery to take TextQuery instead of String as I found I needed to know the start/end index of the field. I added a getter to keep them backwards compatible.

I will look the code later, but if this is a change that could possibly break more than one API, maybe it is worth doing it by bumping the major version and documenting what has been changed in a breaking way. Not yet sure if it is the case here, I'm just stating that it is not off the table.

North101 commented:
If you're happy with a major version bump then that'd make things easier.

isoos commented:
> If you're happy with a major version bump then that'd make things easier.

I don't mind, just make sure to add plenty of notes in the changelog to help users migrate.

isoos commented on lib/src/ast.dart:
```diff
@@ -9,15 +17,18 @@ abstract class Query {
   final String text;
   final bool isExactMatch;
-  TextQuery(this.text, {this.isExactMatch = false});
+  const TextQuery(this.text, super.startIndex, super.endIndex,
```
Fixed position arguments make sense when the list is short and self-evident, but these extra parameters push it beyond the self-evident range. I think it would be better to make them named parameters + `required`, wdyt?

North101 replied:
That's cool with me

isoos commented on lib/src/grammar.dart:
```diff
@@ -199,3 +211,14 @@ class AnyCharExceptPredicate implements CharacterPredicate {
   }
 }
+
+extension on Parser<String> {
```
Nice!

isoos reviewed (commented):
On a high level I like this, but I'd like to do a deep dive into it before going forward with it. Unfortunately I have a few busy days before me, and it may happen only sometime next week :(

North101 commented:
> On a high level I like this, but I'd like to do a deep dive into it before going forward with it. Unfortunately I have a few busy days before me, and it may happen only sometime next week :(

No prob. I'm using this branch in my project for now so no rush

isoos commented on lib/src/ast.dart:
```diff
@@ -27,47 +45,69 @@ class TextQuery implements Query {
-  final String operator;
+class FieldCompareQuery extends Query {
+  final TextQuery field;
+  final TextQuery operator;
```
Is the change from `String` to `TextQuery` for the `operator` field correct?

North101 replied:
It could help figure out where the operator is.

e.g. all of these are valid:
```
field=value
field = value
field    =    value
```

isoos replied:
I think this should be a separate `FieldOperator extends Query` type, with `text` or `value` + `position`. The `TextQuery` has `isExactMatch` which doesn't make sense for this field, and it doesn't feel right. wdyt?

North101 replied:
Does it make sense to have `FieldCompareQuery.field`, `FieldCompareQuery .operator` and `FieldScope.operator` all share the same class. e.g. something like

```dart
class FieldElementQuery extends Query {
  final String text;
  const FieldElementQuery({
    required this.text,
    required super.position,
  });

  @override
  String toString({bool debug = false}) =>
      _debug(debug, text);
}
```

North101 replied:
Or Alternatively

```dart
class SimpleTextQuery extends Query {
  final String text;
  const SimpleTextQuery({
    required this.text,
    required super.position,
  });

  @override
  String toString({bool debug = false}) => _debug(debug, text);
}

/// Text query to match [text].
///
/// [isExactMatch] is set when the [text] was inside quotes.
class TextQuery extends SimpleTextQuery {
```

North101 replied:
Or... Remove `isExactMatch` from TextQuery as its only ever used for `PhraseQuery`?

I think this is what I prefer but I'm happy to for whatever you suggest

isoos replied:
That is a good solution too!

isoos commented on lib/src/ast.dart:
```diff
@@ -1,5 +1,16 @@
 /// Base interface for queries.
 abstract class Query {
+  const Query({
+    required this.startIndex,
```
Having thought of it, I think I'd rather introduce a `SourcePosition` or `SourceSpan` class with `start` and `end` fields, and only have either `position`, `sourcePosition`, `span` or `sourceSpan` field here.
Kind of like in https://pub.dev/documentation/source_span/latest/source_span/SourceSpan-class.html, but less complex. wdyt?

isoos commented on lib/src/ast.dart:
```diff
@@ -9,15 +20,22 @@ abstract class Query {
   @override
   String toString({bool debug = false});
+
+  R cast<R extends Query>() => this as R;
```
Please add a line of documentation, something could be copy-pasted from https://api.dart.dev/stable/2.17.6/dart-core/List/cast.html

isoos reviewed (commented):
Thank you for all the changes you've done! Only one big question remains about the type of the `operator`.

For the other two, I'm happy to do them, but I'm unsure which name to use for the span field/class. Any preference?

isoos commented on lib/src/ast.dart:
```diff
@@ -1,15 +1,25 @@
+  /// The start position of this query, exclusive.
+  final int start;
+
+  /// The end position of this query, exclusive.
```
Isn't this inclusive instead?

North101 replied:
To be honest, i'm not sure. I copied it from https://pub.dev/documentation/source_span/latest/source_span/SourceSpan-class.html

isoos replied:
Ah, sorry, I put the comment at the wrong code line: I meant to ask whether the `start` is inclusive, instead of the current documention stating exclusive.

North101 replied:
Fixed

isoos reviewed (commented):
This looks good! One small nit for the documentation part.

isoos commented:
I'm really happy with the current state of the PR, thank you for idea and all the small follow-ups!
Do you have anything else in the works, or shall I release it?

North101 commented:
That's it for me. Thanks for the library! I've been able to make an advanced SQL db search with it and with this PR it now gives contextual suggests!
![Screenshot_20220629-165430](https://user-images.githubusercontent.com/1031451/180417381-a2049da3-429c-4048-90ba-625546ad7a74.png)
![Screenshot_20220629-165450](https://user-images.githubusercontent.com/1031451/180417389-32b7d39e-a277-4065-91b7-794f3907a346.png)
![Screenshot_20220629-165701](https://user-images.githubusercontent.com/1031451/180417390-01d69b03-14c4-4c15-baf8-b4d10b71fc83.png)

isoos commented:
Great features! :)

Published as 2.0.0
